subprojects/core/src/main/java/org/gradle/internal/classpath/Instrumented.java

```diff
@@ -316,4 +463,232 @@ public Object call(Object receiver, Object arg1) throws Throwable {
             return super.call(receiver, arg1);
         }
     }
+
+    /**
+     * The call site for {@code Runtime.exec}.
+     */
+    private static class ExecCallSite extends AbstractCallSite {
+        public ExecCallSite(CallSite prev) {
+            super(prev);
+        }
+
+        @Override
+        public Object call(Object receiver, Object arg1) throws Throwable {
+            if (checkSignature(receiver, arg1, null, null)) {
+                return callExec(receiver, arg1, null, null);
+            }
+            return super.call(receiver, arg1);
+        }
+
+        @Override
+        public Object call(Object receiver, Object arg1, Object arg2) throws Throwable {
+            if (checkSignature(receiver, arg1, arg2, null)) {
+                return callExec(receiver, arg1, arg2, null);
+            }
+            return super.call(receiver, arg1, arg2);
+        }
+
+        @Override
+        public Object call(Object receiver, Object arg1, Object arg2, Object arg3) throws Throwable {
+            if (checkSignature(receiver, arg1, arg2, arg3)) {
+                return callExec(receiver, arg1, arg2, arg3);
+            }
+            return super.call(receiver, arg1, arg2, arg3);
+        }
+
+        private boolean checkSignature(@Nullable Object runtimeArg, @Nullable Object commandArg, @Nullable Object envpArg, @Nullable Object fileArg) {
```

Comment:
💄 Merge _checkSignature_ and _callExec_ by having the new method return an `Optional<Process?>` which would use `empty` to indicate there was no match?

Reply:
Alternatively, `checkSignature` could return a lambda with the "compiled" implementation or `null` to indicate there was no match.

Either strategy would allow reusing the work already done to find the right match and keep the matching logic and calling logic in sync.